src/osd/ReplicatedPG.cc

```diff
@@ -8311,6 +8359,28 @@ class C_OSD_RepopCommit : public Context {
   }
 };
 
+class C_OSD_RepopCommit_no_pg_lock : public Context {
+  ReplicatedPGRef pg;
+  boost::intrusive_ptr<ReplicatedPG::RepGather> repop;
+  ReplicatedPG::RepGather * real_repop;
+public:
+  C_OSD_RepopCommit_no_pg_lock(ReplicatedPG *pg, ReplicatedPG::RepGather *repop)
+    : pg(pg), repop(repop), real_repop(repop) {}
+  void finish(int flag) {
+    //generic_dout(0) << __func__ << " flag is " << flag << dendl;
+    if (flag == OP_COMP_ACK) {
+      pg->repop_all_committed_no_pg_lock(repop.get());
```

Comment:
@asheplyakov  This is the same as existing C_OSD_RepopApplied (which is the PG lock version).
put() will be called in eval_repop()